Ui: Make the ResultDisplay red on command failure

yamgent opened the pull request:
Part of #784. Rebased, and added tests to verify the result display.

Propsed commit message:
```
ResultDisplay shows an appropriate error message when the user made a 
mistake in his command. The color of the error message is the same as 
the color of a successful one.

Using a different color for error messages makes it easier for a user
to realise that it is communicating the mistake made by him.

Let's make the ResultDisplay red on command failure.
```

_Note:_ For demo purpose, don't merge this to master!

Zhiyuan-Amos commented on src/test/java/seedu/address/ui/CommandBoxTest.java:
```diff
@@ -127,22 +135,38 @@ public void handleKeyPress_startingWithDown() {
         commandBoxHandle.run(COMMAND_THAT_FAILS);
+
+        assertTrue(eventsCollectorRule.eventsCollector.getMostRecent() instanceof NewResultAvailableEvent);
+        assertFalse(((NewResultAvailableEvent) eventsCollectorRule.eventsCollector.getMostRecent()).isSuccessful);
```
Can we remove line 147, since we already have line 148? As per [SelectCommandTest](https://github.com/se-edu/addressbook-level4/blob/master/src/test/java/seedu/address/logic/commands/SelectCommandTest.java#L113), we just cast it without checking as well ^^

If you're updating here, rmb to update below as well.

Zhiyuan-Amos commented on src/test/java/seedu/address/ui/CommandBoxTest.java:
```diff
@@ -127,22 +135,38 @@ public void handleKeyPress_startingWithDown() {
      *      - the command box's style is the same as {@code errorStyleOfCommandBox}.
      */
     private void assertBehaviorForFailedCommand() {
+        eventsCollectorRule.eventsCollector.reset();
```
maybe we should clear after the verifications requiring `eventsCollectorRule#eventsCollector` have completed i.e. at line 150, for the purpose of segmenting `eventsCollector` related things together.

Zhiyuan-Amos commented on src/main/java/seedu/address/ui/ResultDisplay.java:
```diff
@@ -35,7 +38,35 @@ public ResultDisplay() {
+        Platform.runLater(() -> {
+            displayed.setValue(event.message);
+
+            if (event.isSuccessful) {
```
can we ternary this? :P

yamgent replied:
Ternary operators are meant for assigning values.

Zhiyuan-Amos commented on src/main/java/seedu/address/ui/ResultDisplay.java:
```diff
@@ -35,7 +38,35 @@ public ResultDisplay() {
+    /**
+     * Sets the {@code ResultDisplay} style to use the default style.
+     */
+    private void setStyleToDefault() {
```
since we have `setStyleToIndicateCommandFailure`, it seems that this method should be named `setStyleToIndicateCommandSuccess`?

Zhiyuan-Amos commented on src/test/java/guitests/guihandles/ResultDisplayHandle.java:
```diff
@@ -19,4 +20,11 @@ public ResultDisplayHandle(TextArea resultDisplayNode) {
     }
+
+    /**
+     * Returns the list of style classes present in the command box.
```
~~command box~~ result display :P

Zhiyuan-Amos commented on src/test/java/guitests/guihandles/ResultDisplayHandle.java:
```diff
@@ -19,4 +20,11 @@ public ResultDisplayHandle(TextArea resultDisplayNode) {
+    /**
+     * Returns the list of style classes present in the command box.
+     */
+    public ObservableList<String> getStyleClass() {
```
I guess we can return this simply as a `List<String>`, instead of an `Observable<List>`, since we aren't observing it.

Zhiyuan-Amos commented on src/test/java/seedu/address/ui/ResultDisplayTest.java:
```diff
@@ -11,8 +14,14 @@
+            new NewResultAvailableEvent("failure", false);
+
+    private ArrayList<String> defaultStyleOfResultDisplay;
+    private ArrayList<String> errorStyleOfResultDisplay;
```
can these 2 simply be `List<String>`?

Zhiyuan-Amos commented on src/test/java/seedu/address/ui/ResultDisplayTest.java:
```diff
@@ -23,17 +32,38 @@ public void setUp() {
+
+    /**
+     * Posts the {@code event} to the {@code EventsCenter}, then verifies that <br>
+     *      - the text matches the event message <br>
```
the text on the result display matches the {@code event}'s message

Zhiyuan-Amos commented on src/test/java/seedu/address/ui/ResultDisplayTest.java:
```diff
@@ -23,17 +32,38 @@ public void setUp() {
+    /**
+     * Posts the {@code event} to the {@code EventsCenter}, then verifies that <br>
+     *      - the text matches the event message <br>
+     *      - the result display's style is the same as {@code defaultStyleOfResultDisplay} for successful message,
```
`for successful message` sounds weird :P

Zhiyuan-Amos commented on src/test/java/systemtests/AddressBookSystemTest.java:
```diff
@@ -54,6 +55,11 @@
+    private static final List<String> RESULT_DISPLAY_DEFAULT_STYLE =
+            Arrays.asList("text-input", "text-area", "result-display");
+    private static final List<String> RESULT_DISPLAY_ERROR_STYLE =
+            Arrays.asList("text-input", "text-area", "result-display", ResultDisplay.ERROR_STYLE_CLASS);
```
I realise that we have inconsistencies in our code base; for `ResultDisplayTest`, we use `private ArrayList<String> defaultStyleOfResultDisplay` and then we call `defaultStyleOfResultDisplay = new ArrayList<>(resultDisplayHandle.getStyleClass())` to add the style classes. :P

damithc commented:
Just a general comment: after adding tests, this PR is a lot more complicated than before although it is a small UI change. This might overwhelm students. Is it possible to split into even smaller commits so that students can learn one commit at a time.  e.g., add functionality first and then add the tests ?

yamgent commented:
> Is it possible to split into even smaller commits so that students can learn one commit at a time. e.g., add functionality first and then add the tests ?

I split up the commits, so now the unit tests are in [v3 2/5] and [v3 4/5] instead.

yamgent commented:
Changelog:

* fc363ad447bc2afb556e3e1241191ba4e040a476: Updated `HelpCommandSystemTest.java`, because it now uses `assertCommandBoxShowsDefaultStyle()`, but in this PR we renamed this method to `assertCommandBoxAndResultDisplayShowsDefaultStyle()`.
